Draw individual detections and load to Label Studio

VWJF opened the pull request:
- [x] Add detection id text associated with each model detection
  - Using sequential detection ids
  - Text is drawn after rescaling (_see point about image reduction below_)
- [x] (1) draw tile borders for the last row and column and (2) the outer perimeter of the entire image
  - Draw and write text using Pillow
  - Coordinates are calculated manually (using the upper left corner, `0,0` as reference origin) instead of programatically using shapely
- [x] Reduce image before drawing detections/annotations, etc
  - Rearrange order of operations and support rescaling raw coordinated to scaled domain
  - Text is drawn after rescaling
  - [ ] The reduction approach is based on rescaling `TILE_SIZE`; uncertain about the correctness of this approach
  - New order: _Read Image -> Reduce image size -> Collect Boxes -> Draw boxes ... etc_
  - Annotations in both orders are supported by drawing boxes with parametrized outline `width = brush_size.small | brush_size.medium | brush_size.large`
- [x] Ensure/confirm `draw_ground_truth_annotations()` remains unaffected
- [x] Ensure/confirm `draw_mask()` remains unaffected
- [x] Refactor code along the way
  - Extract methods, parametrize constants
  - Add comments, TODOs and flag unused code for removal
- [x] Confirm that code still runs in the `tf-cormorants` venv on Cedar
- [x] Check/address existing TODOs
- [ ] Final refactoring / code cleanup (e.g. auto reformat code, code linter, reorder functions so `def main()` is closer to `if __name__ == '__main__'`)

VWJF committed:
add misc documentation for clarity

VWJF committed:
draw_tiles last row and column were missed and draw outer perimeter

VWJF committed:
refactor draw_detections drawing function

VWJF committed:
assign sequential detection ids and track them while drawing detection boxes

VWJF committed:
draw text in a box attached to the upper left corner of a detection

VWJF committed:
Validate function arguments for rescaling in 1D and 2D

VWJF committed:
update drawing ground truth accordingly --untested

VWJF committed:
After rescaling, the yielded conversions may be float with .0 and need to be cast back to int

VWJF committed:
reorder drawing operations (reduce image size earlier and scale further coordinates): Read Image -> reduce image size -> Collect Boxes -> ...

VWJF committed:
drawing operations, parametrize brush stroke size

VWJF committed:
draw_detections function, add todos and comments on potentially unused functions

VWJF committed:
post_process input validation for missing detections

VWJF committed:
TODO: Calculate text box coordinates programatically using shapely geometric objects

VWJF committed:
refactor to make arguments explicit and undetectable bug fix

jillianderson8 reviewed (commented):
* Agree with removing unused functions. 
* Might need to explicitly specify what version of Python (and some packages) that should be used to run this

VWJF committed:
Update object_detection_scripts/4_comparing_manual_counts/draw_final_detections/src/draw_detections.py

Refactor to remove unused code

Co-authored-by: Jillian Anderson <12687428+jillianderson8@users.noreply.github.com>

VWJF committed:
Update draw_detections.py

Refactor, removing unused code

VWJF committed:
Update draw_detections.py to ensure python3.8 compatibility (type hinting and f-strings with =)

VWJF committed:
Update draw_detections.py to ensure backwards compatibility with python3.7 Pillow

VWJF committed:
Update draw_detections.py
Refactoring for code efficiency

VWJF committed:
Update draw_detections.py
Refactor rescaling argument validation

VWJF committed:
Update draw_detections.py, refactor variable names and parametrize hardcoded values

VWJF committed:
Update draw_detections.py, refactor remove duplicated and unused function standardize_lost()

VWJF committed:
Refactor documentation

VWJF committed:
Update draw_detections.py, update for readbility of output messages

VWJF committed:
Update draw_detections.py, draw individual detections

VWJF committed:
Update draw_detections.py, improving readability of print messages

VWJF committed:
Update draw_detections.py, improve readability of print messages

VWJF committed:
Update draw_detections.py, filter drawing of individual detections by classes

VWJF committed:
Update draw_detections.py, track the individual detection files that have been created

VWJF committed:
Update draw_detections.py. convert image to jpg ('RGB'), export images as png

- to convert to jpg, the image can be converted to 'RGB' format from tiff
- downscaling images and fonts are more readable in png, which is the default output format

VWJF committed:
Update draw_detections.py, export individual detections as csv

VWJF committed:
Update draw_detections.py, fix dangling reference to input file name instead of output file name/format

VWJF committed:
Update draw_detections.py, update call so that rescaling takes place, it was missed earlier

VWJF committed:
Update draw_detections.py, reorder functions for readability

VWJF committed:
Create 4.../label_studio.py, Reshape input csv into a Label Studio task to compare individual nests in day X and a panorama in day Y.

The generated .json can be uploaded to Label Studio to create the annotation task

VWJF committed:
Update draw_detections.py, validate individual_class None type

VWJF committed:
Update draw_detections.py, update progress message for clarity

VWJF committed:
Update draw_detections.py, add cmd argument and switch to handle selection of individual detections

VWJF committed:
Create 1.../label_studio_parsing.ipynb, notebook to preprocess Label Studio annotations and mask into LOST format for the rest of the pipeline

VWJF committed:
Update draw_detections.py, accommodate argparser argument when python version <3.9

Update draw_detections.py, accommodate argparser argument when python version <3.9

VWJF committed:
Update draw_detections.py, misc refactoring and typo fix

VWJF committed:
Update draw_detections.py, accommodate argparser argument when python version <3.9

VWJF committed:
Update draw_detections.py, accommodate argparser argument when python version <3.9

VWJF committed:
Update label_studio.py, change comparison references with 'left', 'right' nomenclature. Implement a cmd arg for swapping left-right column names

VWJF committed:
Update draw_detections.py, draw_box refactored to parametrize optional width argument

VWJF committed:
Update draw_detections.py, imported load_mask from post_process.py to load masks from a list of masks

VWJF committed:
Update draw_detections.py, Improve handling when masks are not found

VWJF committed:
Update draw_detections.py. Update draw_mask() to handle mask loading externally. Improve error handling

VWJF committed:
Update draw_detections.py. Update call to draw_mask()

VWJF committed:
Update draw_detections.py, ImageDraw.polygon functionality depends on pillow_version. Replace with custom function for backwards compatibility

VWJF committed:
Update draw_detections.py, refactoring improve spacing alignment in print

VWJF committed:
Update draw_detections.py, refactor color choices to parametrized dictionary

VWJF committed:
Update draw_detections.py, refactor, update print messaged for clarity

VWJF committed:
Update draw_detections.py, Draw boxes and text with transparency.

- Transparency is implemented using Pillow functionality
- Must convert image to "RGBA" mode
- Transparent colors are selected using 8 digit hex values #rrggbbaa
- It is not as simple as color choice, the transparent objects must be drawn in a new image canvas (the same size as the original) and then combined using `Image.alpha_composite(base, new_canvas)`

https://pillow.readthedocs.io/en/stable/reference/ImageDraw.html#example-draw-partial-opacity-text

VWJF committed:
Update draw_detections.py, remove ground_truth_file and tile_directory input file validation.

Checking for valid paths occurs automatically with pathlib.Path or when reading/writing. No need to check manually

VWJF committed:
Update draw_detections.py, WIP, Load fonts in backwards compatible way using a custom function get_font()

VWJF committed:
Update draw_detections.py, draw the detection_id text under the detection box

VWJF committed:
Update draw_detections.py, improving status print statements

VWJF committed:
Update draw_detections.py. adding lost anchor parameter

VWJF committed:
Update draw_detections.py, WIP-Complete, Load fints in backwards compatible manner with get_font and test_font

VWJF committed:
Update draw_detections.py, partially apply python linter

VWJF committed:
Update draw_detections.py, refactoring for aesthetics improvements

VWJF committed:
Update run_model_on_new_pano.sh, typo causing unintended behavior

VWJF committed:
Update label_studio.py; Path().with_stem() is not supported in python3.7, replace with with_name() and with_suffix() for backwards compatibility

VWJF committed:
Update label_studio.py; extract pano name from individual detections file and misc refactoring

VWJF committed:
Update label_studio.py; dropna files instead of keeping them. Extract relative path to individual file programmatically.

VWJF committed:
Update label_studio.py; refactor: remove unused code

VWJF committed:
Update label_studio.py; function input to match type

VWJF committed:
Update label_studio.py; Remove restriction of df.head(2) meant for development purposes

VWJF committed:
Update shell driver scripts for drawing individual detections and label studio parsing

VWJF committed:
Update execute permissions of shell files

VWJF committed:
Update label_studio.py; refactor: improve docstring

VWJF committed:
Update draw_detections.py; Address todo about replacing an input validation function with type hints. Due to differences between python 3.7 and 3.8, it is prudent to use an explicit function. Explicit is better than implicit.

VWJF committed:
Update draw_detections.py; Address Todo about code duplication, the duplication concern was addressed with parametrized anno_tile size, and other tile sizes. Removing todo.

VWJF committed:
Update draw_detections.py; Add example usage
